README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# comorbidity <img src="man/figures/hex.png" width = "150" align="right" />

2019-12-26

[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/ellessenne/comorbidity?branch=master&svg=true)](https://ci.appveyor.com/project/ellessenne/comorbidity)
[![Travis-CI Build
Status](https://travis-ci.org/ellessenne/comorbidity.svg?branch=master)](https://travis-ci.org/ellessenne/comorbidity)
[![Coverage
Status](https://img.shields.io/codecov/c/github/ellessenne/comorbidity/master.svg)](https://codecov.io/github/ellessenne/comorbidity?branch=master)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/comorbidity)](https://cran.r-project.org/package=comorbidity)
[![CRAN\_Logs\_Badge](http://cranlogs.r-pkg.org/badges/comorbidity)](https://cran.r-project.org/package=comorbidity)
[![CRAN\_Logs\_Badge\_Total](http://cranlogs.r-pkg.org/badges/grand-total/comorbidity)](https://cran.r-project.org/package=comorbidity)
[![JOSS
DOI](http://joss.theoj.org/papers/10.21105/joss.00648/status.svg)](https://doi.org/10.21105/joss.00648)
[![Zenodo
DOI](https://zenodo.org/badge/68221970.svg)](https://zenodo.org/badge/latestdoi/68221970)
[![PRs
Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](http://makeapullrequest.com)

`comorbidity` is an R package for computing comorbidity scores such as
the weighted Charlson score and the Elixhauser comorbidity score; both
ICD-10 and ICD-9 coding systems are supported.

## Installation

`comorbidity` is on CRAN. You can install it as usual with:

``` r
install.packages("comorbidity")
```

Alternatively, you can install the development version from GitHub with:

``` r
# install.packages("remotes")
remotes::install_github("ellessenne/comorbidity")
```

## Simulating ICD-10 codes

The `comorbidity` packages includes a function named `sample_diag()`
that allows simulating ICD diagnostic codes in a straightforward way.
For instance, we could simulate ICD-10 codes:

``` r
# load the comorbidity package
library(comorbidity)
# set a seed for reproducibility
set.seed(1)
# simulate 50 ICD-10 codes for 5 individuals
x <- data.frame(
  id = sample(1:5, size = 50, replace = TRUE),
  code = sample_diag(n = 50),
  stringsAsFactors = FALSE
)
x <- x[order(x$id, x$code), ]
print(head(x, n = 15), row.names = FALSE)
##  id code
##   1  B02
##   1 B582
##   1 I749
##   1 J450
##   1 L893
##   1 Q113
##   1  Q26
##   1 Q978
##   1 T224
##   1 V101
##   1 V244
##   1  V46
##   2 A665
##   2 C843
##   2 D838
```

It is also possible to simulate from two different versions of the
ICD-10 coding system. The default is to simulate ICD-10 codes from the
2011 version:

``` r
set.seed(1)
x1 <- data.frame(
  id = sample(1:3, size = 30, replace = TRUE),
  code = sample_diag(n = 30),
  stringsAsFactors = FALSE
)
set.seed(1)
x2 <- data.frame(
  id = sample(1:3, size = 30, replace = TRUE),
  code = sample_diag(n = 30, version = "ICD10_2011"),
  stringsAsFactors = FALSE
)
# should return TRUE
all.equal(x1, x2)
## [1] TRUE
```

Alternatively, you could use the 2009 version:

``` r
set.seed(1)
x1 <- data.frame(
  id = sample(1:3, size = 30, replace = TRUE),
  code = sample_diag(n = 30, version = "ICD10_2009"),
  stringsAsFactors = FALSE
)
set.seed(1)
x2 <- data.frame(
  id = sample(1:3, size = 30, replace = TRUE),
  code = sample_diag(n = 30, version = "ICD10_2011"),
  stringsAsFactors = FALSE
)
# should not return TRUE
all.equal(x1, x2)
## [1] "Component \"code\": 30 string mismatches"
```

## Simulating ICD-9 codes

ICD-9 codes can be easily simulated too:

``` r
set.seed(2)
x9 <- data.frame(
  id = sample(1:3, size = 30, replace = TRUE),
  code = sample_diag(n = 30, version = "ICD9_2015"),
  stringsAsFactors = FALSE
)
x9 <- x9[order(x9$id, x9$code), ]
print(head(x9, n = 15), row.names = FALSE)
##  id  code
##   1 01130
##   1 01780
##   1 30151
##   1  3073
##   1 36907
##   1 37845
##   1 64212
##   1 66704
##   1 72633
##   1  9689
##   1  V289
##   2  0502
##   2 09169
##   2 20046
##   2 25082
```

## Computing comorbidity scores

The main function of the `comorbidity` package is named `comorbidity()`,
and it can be used to compute any supported comorbidity score; scores
can be specified by setting the `score` argument, which is required.

Say we have 3 individuals with a total of 30 ICD-10 diagnostic codes:

``` r
set.seed(1)
x <- data.frame(
  id = sample(1:3, size = 30, replace = TRUE),
  code = sample_diag(n = 30),
  stringsAsFactors = FALSE
)
```

We could compute the Charlson score, index, and each comorbidity domain:

``` r
charlson <- comorbidity(x = x, id = "id", code = "code", score = "charlson", icd = "icd10", assign0 = FALSE)
charlson
##   id ami chf pvd cevd dementia copd rheumd pud mld diab diabwc hp rend canc msld metacanc aids
## 1  1   0   0   0    0        0    0      0   0   0    0      0  0    0    1    0        0    1
## 2  2   0   0   0    0        0    0      0   0   0    0      0  0    0    1    0        0    0
## 3  3   0   0   0    0        0    0      0   0   0    0      0  0    0    0    0        0    0
##   score index wscore windex
## 1     2   1-2      8    >=5
## 2     1   1-2      2    1-2
## 3     0     0      0      0
```

We set the `assign0` argument to `FALSE` to not apply a hierarchy of
comorbidity codes, as described in `?comorbidity::comorbidity`. The
default is to assume ICD-10 codes are passed to `comorbidity`:

``` r
charlson.default <- comorbidity(x = x, id = "id", code = "code", score = "charlson", assign0 = FALSE)
all.equal(charlson, charlson.default)
## [1] TRUE
```

Alternatively, we could compute the Elixhauser score:

``` r
elixhauser <- comorbidity(x = x, id = "id", code = "code", score = "elixhauser", icd = "icd10", assign0 = FALSE)
elixhauser
##   id chf carit valv pcd pvd hypunc hypc para ond cpd diabunc diabc hypothy rf ld pud aids lymph
## 1  1   0     0    0   0   0      0    0    0   0   0       0     0       0  0  0   0    1     0
## 2  2   0     0    1   0   0      0    0    0   0   0       0     0       0  0  0   0    0     0
## 3  3   0     0    0   0   0      0    0    0   1   0       0     0       0  0  0   0    0     0
##   metacanc solidtum rheumd coag obes wloss fed blane dane alcohol drug psycho depre score index
## 1        0        1      0    0    0     0   0     0    0       0    0      0     0     2   1-4
## 2        0        1      0    0    0     0   0     0    0       0    0      0     0     2   1-4
## 3        0        0      0    0    0     0   0     0    0       0    0      0     0     1   1-4
##   wscore_ahrq wscore_vw windex_ahrq windex_vw
## 1           7         4         >=5       1-4
## 2           7         3         >=5       1-4
## 3           5         6         >=5       >=5
```

Conversely, say we have 5 individuals with a total of 100 ICD-9
diagnostic codes:

``` r
set.seed(3)
x <- data.frame(
  id = sample(1:5, size = 100, replace = TRUE),
  code = sample_diag(n = 100, version = "ICD9_2015"),
  stringsAsFactors = FALSE
)
```

The Charlson and Elixhauser comorbidity codes can be easily computed:

We could compute the Charlson score, index, and each comorbidity domain:

``` r
charlson9 <- comorbidity(x = x, id = "id", code = "code", score = "charlson", icd = "icd9", assign0 = FALSE)
charlson9
##   id ami chf pvd cevd dementia copd rheumd pud mld diab diabwc hp rend canc msld metacanc aids
## 1  1   0   0   1    0        0    0      0   0   0    0      0  0    0    1    0        0    0
## 2  2   0   0   0    1        0    0      0   0   0    0      0  0    0    0    0        0    0
## 3  3   0   0   0    0        0    0      0   1   0    0      0  0    0    0    0        0    0
## 4  4   0   0   1    1        0    0      0   0   0    0      0  0    0    1    0        0    0
## 5  5   0   0   0    0        0    0      0   0   0    0      0  0    0    1    0        0    0
##   score index wscore windex
## 1     2   1-2      3    3-4
## 2     1   1-2      1    1-2
## 3     1   1-2      1    1-2
## 4     3   3-4      4    3-4
## 5     1   1-2      2    1-2
```

Alternatively, we could compute the Elixhauser score:

``` r
elixhauser9 <- comorbidity(x = x, id = "id", code = "code", score = "elixhauser", icd = "icd9", assign0 = FALSE)
elixhauser9
##   id chf carit valv pcd pvd hypunc hypc para ond cpd diabunc diabc hypothy rf ld pud aids lymph
## 1  1   0     0    0   0   1      0    0    0   0   0       0     0       0  0  0   0    0     0
## 2  2   0     0    0   0   0      0    0    0   1   0       0     0       0  0  0   0    0     0
## 3  3   0     0    0   0   0      0    0    0   0   0       0     0       0  0  0   0    0     0
## 4  4   0     0    0   1   1      0    0    0   0   0       0     0       0  0  0   0    0     0
## 5  5   0     0    0   0   0      0    0    0   0   0       0     0       0  0  0   0    0     0
##   metacanc solidtum rheumd coag obes wloss fed blane dane alcohol drug psycho depre score index
## 1        0        0      0    0    0     0   0     0    0       0    0      0     0     1   1-4
## 2        0        0      0    0    0     0   0     0    0       0    0      0     0     1   1-4
## 3        0        0      0    0    0     0   0     0    0       0    0      1     0     1   1-4
## 4        0        0      0    0    0     0   0     0    0       0    0      0     0     2   1-4
## 5        0        0      1    0    0     0   0     0    0       0    0      0     0     1   1-4
##   wscore_ahrq wscore_vw windex_ahrq windex_vw
## 1           3         2         1-4       1-4
## 2           5         6         >=5       >=5
## 3          -5         0          <0         0
## 4           9         6         >=5       >=5
## 5           0         0           0         0
```

The weighted Elixhauser score is computed using both the AHRQ and the
van Walraven algorithm (`wscore_ahrq` and `wscore_vw`).

## Citation

If you find `comorbidity` useful, please cite it in your publications:

``` r
citation("comorbidity")
## 
## To cite the comorbidity package in publications, please use:
## 
##   Gasparini, (2018). comorbidity: An R package for computing comorbidity scores. Journal
##   of Open Source Software, 3(23), 648, https://doi.org/10.21105/joss.00648
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     author = {Alessandro Gasparini},
##     title = {comorbidity: An R package for computing comorbidity scores},
##     journal = {Journal of Open Source Software},
##     year = {2018},
##     volume = {3},
##     issue = {23},
##     pages = {648},
##     doi = {10.21105/joss.00648},
##     url = {https://doi.org/10.21105/joss.00648},
##   }
```

## References

This package is based on the ICD-10-based formulations of the Charlson
score and Elixhauser score proposed by Quan *et al*. in 2005. The ICD-9
formulation of the Charlson score is also from Quan *et al*. The
ICD-9-based Elixhauser score is according to the AHRQ formulation (Moore
*et al*., 2017). Weights for the Charlson score are based on the
original formulation by Charlson *et al*. in 1987, while weights for the
Elixhauser score are based on work by van Walraven *et al*. Finally, the
categorisation of scores and weighted scores is based on work by
Menendez *et al*. Further details on each algorithm are included in the
package vignette, which you can access by typing the following in the R
console:

``` r
vignette("comorbidityscores", package = "comorbidity")
```

  - Quan H, Sundararajan V, Halfon P, Fong A, Burnand B, Luthi JC, et
    al. *Coding algorithms for defining comorbidities in ICD-9-CM and
    ICD-10 administrative data*. Medical Care 2005; 43(11):1130-1139.
    DOI:
    [10.1097/01.mlr.0000182534.19832.83](https://doi.org/10.1097/01.mlr.0000182534.19832.83)
  - Charlson ME, Pompei P, Ales KL, et al. *A new method of classifying
    prognostic comorbidity in longitudinal studies: development and
    validation*. Journal of Chronic Diseases 1987; 40:373-383. DOI:
    [10.1016/0021-9681(87)90171-8](https://doi.org/10.1016/0021-9681\(87\)90171-8)
  - Moore BJ, White S, Washington R, Coenen N, and Elixhauser A.
    *Identifying increased risk of readmission and in-hospital mortality
    using hospital administrative data: the AHRQ Elixhauser comorbidity
    index*. Medical Care 2017; 55(7):698-705. DOI:
    [10.1097/MLR.0000000000000735](https://doi.org/10.1097/MLR.0000000000000735)
  - Elixhauser A, Steiner C, Harris DR and Coffey RM. *Comorbidity
    measures for use with administrative data*. Medical Care 1998;
    36(1):8-27. DOI:
    [10.1097/00005650-199801000-00004](https://doi.org/10.1097/00005650-199801000-00004)
  - van Walraven C, Austin PC, Jennings A, Quan H and Forster AJ. *A
    modification of the Elixhauser comorbidity measures into a point
    system for hospital death using administrative data*. Medical Care
    2009; 47(6):626-633. DOI:
    [10.1097/mlr.0b013e31819432e5](https://doi.org/10.1097/mlr.0b013e31819432e5)
  - Menendez ME, Neuhaus V, van Dijk CN, Ring D. *The Elixhauser
    comorbidity method outperforms the Charlson index in predicting
    inpatient death after orthopaedic surgery*. Clinical Orthopaedics
    and Related Research 2014; 472(9):2878-2886. DOI:
    [10.1007/s11999-014-3686-7](https://doi.org/10.1007/s11999-014-3686-7)

## Copyright

The icon for the hex sticker was made by
[monkik](https://www.flaticon.com/authors/monkik) from
[www.flaticon.com](https://www.flaticon.com), and is licensed by
[Creative Commons BY 3.0](http://creativecommons.org/licenses/by/3.0).


NEWS.md
# comorbidity 0.5.3

* `comorbidity` now accepts `data.table` objects as input (#23);

* `comorbidity` can deal with non-syntactically valid names (#25);

* Fixed coding typo in vignette (#10, #26);

* Reduced number of random tests, as R CMD check was taking too long on some platforms getting killed on CRAN (as reported by Kurt Hornik).

# comorbidity 0.5.2

* Fixed another bug introduced by the switch to `data.table` that only occurred when the input dataset had extra columns other than `id` and `code`.

# comorbidity 0.5.1

* The `melt` generic in `data.table` has been deprecated; the dependency on `reshape2` has now been removed;

* Fixed bugs introduced by switching to `data.table` in `comorbidity` 0.5.0: column names `id` and `code` were expecting these specific names, now fixed and behaves as expected.

# comorbidity 0.5.0

* `comorbidity` is now 10+ times faster, thanks to [Jonathan Williman](https://github.com/jwilliman) contributing code based on the `data.table` package;

* Fixed typo in vignette regarding weighting algorithm for the AHRQ Elixhauser comorbidity score (#14, thanks to @cornflakegrl);

* Added `pkgdown` website: https://ellessenne.github.io/comorbidity.

# comorbidity 0.4.1

* Fixed bug in regex patterns (#10, thanks to @francisco003 for reporting it and @salmasian for the pull request).

# comorbidity 0.4.0

### BREAKING CHANGES

Modified the behaviour of the `assign0` argument after further discussion with [Anders Alexandersson](https://github.com/aalexandersson) in Issue #9: now there is no default, forcing the user to decide whether to apply a hierarchy of comorbidity codes or not. This will make the algorithm more transparent to the end user, allowing an informed choice. See `?comorbidity::comorbidity` and `vignette("comorbidityscores", package = "comorbidity")` for further details on the hierarchy being applied.

# comorbidity 0.3.0

### BREAKING CHANGES

`comorbidity` now returns two Elixhauser scores, one computed using the algorithm of val Walraven _et al_. (2009) and a second one computed using the AHRQ algorithm (Moore _et al_., 2017). Thanks to Yumiko Abe-Jones for feedback and the discussion regarding weighted Elixhauser scores.

More information can be found on the package vignette: `vignette("comorbidityscores", package = "comorbidity")`.

# comorbidity 0.2.1

* Fixed bug in weighting algorithm of Elixhauser comorbidity score;
* The `assign0` argument of `comorbidity` now defaults to `FALSE`;
* Improved documentation for the `comorbidity` function:
    - `assign0` now explains in details what hierarchy of comorbidities is applied;
    - added reference to package vignette where comorbidity scores and weighting algorithms are explained in more detail.

# comorbidity 0.2.0

* `comorbidity` is faster, with a conservative estimated speed-up of >60%;
* Lots of internal housekeeping;
* Fixed broken GitHub links to the R script used to generate the datasets bundled with `comorbidity`.

### BREAKING CHANGES

The `score` argument from `comorbidity` has been split into `score` and `icd`. For instance, the command `comorbidity(x = x, id = "id", code = "code", score = "charlson_icd10")` has to be modified as `r comorbidity(x = x, id = "id", code = "code", score = "charlson", icd = "icd10")`. The default value of `icd` is `icd10`, for ICD-10 codes, and possible values are `icd10` and `icd9`.

# comorbidity 0.1.3

* Added `nhds2010` and `australia10` datasets, imported from Stata version 15.

Bug fix:
* Fixed a bug in the regex for the ICD10 Charlson score;
* Fixed a bug in the regex for the ICD10 Elixhauser score.

# comorbidity 0.1.2

* Added ICD10-CM data (version 2017 and 2018).

# comorbidity 0.1.1

* Documented variables that were missing among those returned by `comorbidity()` (@corinne-riddell, #5);
* Added CITATION file: `citation("comorbidity")` now returns a properly formatted entry.

# comorbidity 0.1.0

* Added support for the ICD-9-CM version of Charlson and Elixhauser scores
* Added vignette with information on the scores computed by `comorbidity`
* `sample_diag_icd10()` function renamed back to `sample_diag()`, as now can simulate ICD-9-CM codes too

# comorbidity 0.0.3

* Added datasets `icd10_2009` and `icd10_2011` with ICD-10 codes, 2009 and 2011 versions (respectively)
* `sample_diag` is now `sample_diag_icd10` and simulates proper ICD-10 codes
* Added Elixhauser comorbidity score
* Renamed the package to `comorbidity` as it now can compute more than just the Charlson score
* Added formal testing: codes for the Charlson score are properly identified
* Added formal testing: codes for the Elixhauser score are properly identified

# charlson 0.0.2

* Running computations in parallel now should work on every platform
* Improved code coverage
* `charlson` is marginally faster

# charlson 0.0.1

* Rebooted `charlson` using only base R functions
* Added a `NEWS.md` file to track changes to the package
* Added CI with Travis and AppVeyor
* Added automated testing with `testthat`
* Added code coverage with Codecov


R/comorbidity.R
#' @title Compute comorbidity scores.
#'
#' @description Computes comorbidity scores such as the weighted Charlson score and the Elixhauser comorbidity score.
#'
#' @param x A tidy `data.frame` (or a `data.table`; `tibble`s are supported too) with one column containing an individual ID and a column containing all diagnostic codes.
#' Extra columns other than ID and codes are discarded.
#' Column names must be syntactically valid names, otherwise they are forced to be so by calling the [make.names()] function.
#' @param id Column of `x` containing the individual ID.
#' @param code Column of `x` containing diagnostic codes. Codes must be in upper case with no punctuation in order to be properly recognised.
#' @param score The comorbidity score to compute. Possible choices are the weighted Charlson score (`charlson`) and the weighted Elixhauser score (`elixhauser`). Values are case-insensitive.
#' @param assign0 Apply a hierarchy of comorbidities. If `TRUE`, should a comorbidity be present in a patient with different degrees of severity, then the milder form will be assigned to 0 and therefore not counted. By doing this, a type of comorbidity is not counted more than once in each patient. In particular, the comorbidities that are affected by this argument are:
#' * "Mild liver disease" (`mld`) and "Moderate/severe liver disease" (`msld`) for the Charlson score;
#' * "Diabetes" (`diab`) and "Diabetes with complications" (`diabwc`) for the Charlson score;
#' * "Cancer" (`canc`) and "Metastatic solid tumour" (`metacanc`) for the Charlson score;
#' * "Hypertension, uncomplicated" (`hypunc`) and "Hypertension, complicated" (`hypc`) for the Elixhauser score;
#' * "Diabetes, uncomplicated" (`diabunc`) and "Diabetes, complicated" (`diabc`) for the Elixhauser score;
#' * "Solid tumour" (`solidtum`) and "Metastatic cancer" (`metacanc`) for the Elixhauser score.
#' @param icd The version of ICD coding to use. Possible choices are ICD-9-CM (`icd9`) or ICD-10 (`icd10`). Defaults to `icd10`, and values are case-insensitive.
#' @param factorise Return comorbidities as factors rather than numeric, where (1 = presence of comorbidity, 0 = otherwise). Defaults to `FALSE`.
#' @param labelled Attach labels to each comorbidity, compatible with the RStudio viewer via the [utils::View()] function. Defaults to `TRUE`.
#' @param tidy.codes Tidy diagnostic codes? If `TRUE`, all codes are converted to upper case and all non-alphanumeric characters are removed using the regular expression \code{[^[:alnum:]]}. Defaults to `TRUE`.
#' @return A data frame with `id`, columns relative to each comorbidity domain, comorbidity score, weighted comorbidity score, and categorisations of such scores, with one row per individual.
#'
#' For the Charlson score, the following variables are included in the dataset:
#' * The `id` variable as defined by the user;
#' * `ami`, for acute myocardial infarction;
#' * `chf`, for congestive heart failure;
#' * `pvd`, for peripheral vascular disease;
#' * `cevd`, for cerebrovascular disease;
#' * `dementia`, for dementia;
#' * `copd`, chronic obstructive pulmonary disease;
#' * `rheumd`, for rheumatoid disease;
#' * `pud`, for peptic ulcer disease;
#' * `mld`, for mild liver disease;
#' * `diab`, for diabetes without complications;
#' * `diabwc`, for diabetes with complications;
#' * `hp`, for hemiplegia or paraplegia;
#' * `rend`, for renal disease;
#' * `canc`, for cancer (any malignancy);
#' * `msld`, for moderate or severe liver disease;
#' * `metacanc`, for metastatic solid tumour;
#' * `aids`, for AIDS/HIV;
#' * `score`, for the non-weighted version of the Charlson score;
#' * `index`, for the non-weighted version of the grouped Charlson index;
#' * `wscore`, for the weighted version of the Charlson score;
#' * `windex`, for the weighted version of the grouped Charlson index.
#'
#' Conversely, for the Elixhauser score the dataset contains the following variables:
#' * The `id` variable as defined by the user;
#' * `chf`, for congestive heart failure;
#' * `carit`, for cardiac arrhythmias;
#' * `valv`, for valvular disease;
#' * `pcd`, for pulmonary circulation disorders;
#' * `pvd`, for peripheral vascular disorders;
#' * `hypunc`, for hypertension, uncomplicated;
#' * `hypc`, for hypertension, complicated;
#' * `para`, for paralysis;
#' * `ond`, for other neurological disorders;
#' * `cpd`, for chronic pulmonary disease;
#' * `diabunc`, for diabetes, uncomplicated;
#' * `diabc`, for diabetes, complicated;
#' * `hypothy`, for hypothyroidism;
#' * `rf`, for renal failure;
#' * `ld`, for liver disease;
#' * `pud`, for peptic ulcer disease, excluding bleeding;
#' * `aids`, for AIDS/HIV;
#' * `lymph`, for lymphoma;
#' * `metacanc`, for metastatic cancer;
#' * `solidtum`, for solid tumour, without metastasis;
#' * `rheumd`, for rheumatoid arthritis/collaged vascular disease;
#' * `coag`, for coagulopathy;
#' * `obes`, for obesity;
#' * `wloss`, for weight loss;
#' * `fed`, for fluid and electrolyte disorders;
#' * `blane`, for blood loss anaemia;
#' * `dane`, for deficiency anaemia;
#' * `alcohol`, for alcohol abuse;
#' * `drug`, for drug abuse;
#' * `psycho`, for psychoses;
#' * `depre`, for depression;
#' * `score`, for the non-weighted version of the Elixhauser score;
#' * `index`, for the non-weighted version of the grouped Elixhauser index;
#' * `wscore_ahrq`, for the weighted version of the Elixhauser score using the AHRQ algorithm (Moore _et al_., 2017);
#' * `wscore_vw`, for the weighted version of the Elixhauser score using the algorithm in van Walraven _et al_. (2009);
#' * `windex_ahrq`, for the weighted version of the grouped Elixhauser index using the AHRQ algorithm (Moore _et al_., 2017);
#' * `windex_vw`, for the weighted version of the grouped Elixhauser index using the algorithm in van Walraven _et al_. (2009).
#'
#' Labels are presented to the user when using the RStudio viewer (e.g. via the [utils::View()] function) for convenience.
#'
#' @details
#' The ICD-10 and ICD-9-CM coding for the Charlson and Elixhauser scores is based on work by Quan _et al_. (2005). Weights for the Charlson score are based on the original formulation by Charlson _et al_. in 1987, while weights for the Elixhauser score are based on work by Moore _et al_. and van Walraven _et al_. Finally, the categorisation of scores and weighted scores is based on work by Menendez _et al_. See `vignette("comorbidityscores", package = "comorbidity")` for further details on the comorbidity scores and the weighting algorithm.
#' ICD-10 and ICD-9 codes must be in upper case and with alphanumeric characters only in order to be properly recognised; set `tidy.codes = TRUE` to properly tidy the codes automatically. As a convenience, a message is printed to the R console when non-alphanumeric characters are found.
#'
#' @references Quan H, Sundararajan V, Halfon P, Fong A, Burnand B, Luthi JC, et al. _Coding algorithms for defining comorbidities in ICD-9-CM and ICD-10 administrative data_. Medical Care 2005; 43(11):1130-1139.
#' @references Charlson ME, Pompei P, Ales KL, et al. _A new method of classifying prognostic comorbidity in longitudinal studies: development and validation_. Journal of Chronic Diseases 1987; 40:373-383.
#' @references Moore BJ, White S, Washington R, Coenen N, and Elixhauser A. _Identifying increased risk of readmission and in-hospital mortality using hospital administrative data: the AHRQ Elixhauser comorbidity index_. Medical Care 2017; 55(7):698-705.
#' @references van Walraven C, Austin PC, Jennings A, Quan H and Forster AJ. _A modification of the Elixhauser comorbidity measures into a point system for hospital death using administrative data_. Medical Care 2009; 47(6):626-633.
#' @references Menendez ME, Neuhaus V, van Dijk CN, Ring D. _The Elixhauser comorbidity method outperforms the Charlson index in predicting inpatient death after orthopaedic surgery_. Clinical Orthopaedics and Related Research 2014; 472(9):2878-2886.
#'
#' @examples
#' set.seed(1)
#' x <- data.frame(
#'   id = sample(1:15, size = 200, replace = TRUE),
#'   code = sample_diag(200),
#'   stringsAsFactors = FALSE
#' )
#'
#' # Charlson score based on ICD-10 diagnostic codes:
#' comorbidity(x = x, id = "id", code = "code", score = "charlson", assign0 = FALSE)
#'
#' # Elixhauser score based on ICD-10 diagnostic codes:
#' comorbidity(x = x, id = "id", code = "code", score = "elixhauser", assign0 = FALSE)
#' @export

comorbidity <- function(x, id, code, score, assign0, icd = "icd10", factorise = FALSE, labelled = TRUE, tidy.codes = TRUE) {
  ### Check arguments
  arg_checks <- checkmate::makeAssertCollection()
  # x must be a data.frame (or a data.table)
  checkmate::assert_true(all(class(x) %in% c("data.frame", "data.table", "tbl", "tbl_df")), add = arg_checks)
  # id, code, score, icd must be a single string value
  checkmate::assert_string(id, add = arg_checks)
  checkmate::assert_string(code, add = arg_checks)
  checkmate::assert_string(score, add = arg_checks)
  checkmate::assert_string(icd, add = arg_checks)
  # score must be charlson, elixhauser; case insensitive
  score <- tolower(score)
  checkmate::assert_choice(score, choices = c("charlson", "elixhauser"), add = arg_checks)
  # icd must be icd9, icd10; case insensitive
  icd <- tolower(icd)
  checkmate::assert_choice(icd, choices = c("icd9", "icd10"), add = arg_checks)
  # assign0, factorise, labelled, tidy.codes, parallel must be a single boolean value
  checkmate::assert_logical(assign0, len = 1, add = arg_checks)
  checkmate::assert_logical(factorise, len = 1, add = arg_checks)
  checkmate::assert_logical(labelled, len = 1, add = arg_checks)
  checkmate::assert_logical(tidy.codes, len = 1, add = arg_checks)
  # force names to be syntactically valid:
  if (any(names(x) != make.names(names(x)))) {
    names(x) <- make.names(names(x))
    warning("Names of the input dataset 'x' have been modified by make.names(). See ?make.names() for more details.", call. = FALSE)
  }
  if (id != make.names(id)) {
    id <- make.names(id)
    warning("The input 'id' string has been modified by make.names(). See ?make.names() for more details.", call. = FALSE)
  }
  if (code != make.names(code)) {
    code <- make.names(code)
    warning("The input 'id' string has been modified by make.names(). See ?make.names() for more details.", call. = FALSE)
  }
  # id, code must be in x
  checkmate::assert_subset(id, choices = names(x), add = arg_checks)
  checkmate::assert_subset(code, choices = names(x), add = arg_checks)
  # Report if there are any errors
  if (!arg_checks$isEmpty()) checkmate::reportAssertions(arg_checks)

  ### Tidy codes if required
  if (tidy.codes) x <- .tidy(x = x, code = code)

  ### Extract regex for internal use
  regex <- lofregex[[score]][[icd]]

  ### Subset only 'id' and 'code' columns
  if (data.table::is.data.table(x)) {
    x <- x[, c(id, code), with = FALSE]
  } else {
    x <- x[, c(id, code)]
  }

  ### Turn x into a DT
  data.table::setDT(x)

  ### Get list of unique codes used in dataset that match comorbidities
  loc <- sapply(regex, grep, unique(x[[code]]), value = TRUE)
  loc <- utils::stack(loc)
  names(loc)[1] <- code

  ### Merge list with original data.table (data.frame)
  x <- merge(x, loc, all.x = TRUE, allow.cartesian = TRUE)
  x[[code]] <- NULL
  x <- unique(x)

  ### Spread wide
  xin <- x[, c(id, "ind"), with = FALSE]
  xin[, value := 1L]
  x <- data.table::dcast.data.table(xin, stats::as.formula(paste(id, "~ ind")), fill = 0)
  x[["NA"]] <- NULL

  ### Add missing columns
  for (col in names(regex)) {
    if (is.null(x[[col]])) x[[col]] <- 0
  }
  data.table::setcolorder(x, c(id, names(regex)))

  ### Turn internal DT into a DF
  data.table::setDF(x)

  ### Compute Charlson score and Charlson index
  if (score == "charlson") {
    x$score <- with(x, ami + chf + pvd + cevd + dementia + copd + rheumd + pud + mld * ifelse(msld == 1 & assign0, 0, 1) + diab * ifelse(diabwc == 1 & assign0, 0, 1) + diabwc + hp + rend + canc * ifelse(metacanc == 1 & assign0, 0, 1) + msld + metacanc + aids)
    x$index <- with(x, cut(score, breaks = c(0, 1, 2.5, 4.5, Inf), labels = c("0", "1-2", "3-4", ">=5"), right = FALSE))
    x$wscore <- with(x, ami + chf + pvd + cevd + dementia + copd + rheumd + pud + mld * ifelse(msld == 1 & assign0, 0, 1) + diab * ifelse(diabwc == 1 & assign0, 0, 1) + diabwc * 2 + hp * 2 + rend * 2 + canc * ifelse(metacanc == 1 & assign0, 0, 2) + msld * 3 + metacanc * 6 + aids * 6)
    x$windex <- with(x, cut(wscore, breaks = c(0, 1, 2.5, 4.5, Inf), labels = c("0", "1-2", "3-4", ">=5"), right = FALSE))
  } else {
    x$score <- with(x, chf + carit + valv + pcd + pvd + hypunc * ifelse(hypc == 1 & assign0, 0, 1) + hypc + para + ond + cpd + diabunc * ifelse(diabc == 1 & assign0, 0, 1) + diabc + hypothy + rf + ld + pud + aids + lymph + metacanc + solidtum * ifelse(metacanc == 1 & assign0, 0, 1) + rheumd + coag + obes + wloss + fed + blane + dane + alcohol + drug + psycho + depre)
    x$index <- with(x, cut(score, breaks = c(-Inf, 0, 1, 4.5, Inf), labels = c("<0", "0", "1-4", ">=5"), right = FALSE))
    x$wscore_ahrq <- with(x, chf * 9 + carit * 0 + valv * 0 + pcd * 6 + pvd * 3 + ifelse(hypunc == 1 | hypc == 1, 1, 0) * (-1) + para * 5 + ond * 5 + cpd * 3 + diabunc * ifelse(diabc == 1 & assign0, 0, 0) + diabc * (-3) + hypothy * 0 + rf * 6 + ld * 4 + pud * 0 + aids * 0 + lymph * 6 + metacanc * 14 + solidtum * ifelse(metacanc == 1 & assign0, 0, 7) + rheumd * 0 + coag * 11 + obes * (-5) + wloss * 9 + fed * 11 + blane * (-3) + dane * (-2) + alcohol * (-1) + drug * (-7) + psycho * (-5) + depre * (-5))
    x$wscore_vw <- with(x, chf * 7 + carit * 5 + valv * (-1) + pcd * 4 + pvd * 2 + ifelse(hypunc == 1 | hypc == 1, 1, 0) * 0 + para * 7 + ond * 6 + cpd * 3 + diabunc * ifelse(diabc == 1 & assign0, 0, 0) + diabc * 0 + hypothy * 0 + rf * 5 + ld * 11 + pud * 0 + aids * 0 + lymph * 9 + metacanc * 12 + solidtum * ifelse(metacanc == 1 & assign0, 0, 4) + rheumd * 0 + coag * 3 + obes * (-4) + wloss * 6 + fed * 5 + blane * (-2) + dane * (-2) + alcohol * 0 + drug * (-7) + psycho * 0 + depre * (-3))
    x$windex_ahrq <- with(x, cut(wscore_ahrq, breaks = c(-Inf, 0, 1, 4.5, Inf), labels = c("<0", "0", "1-4", ">=5"), right = FALSE))
    x$windex_vw <- with(x, cut(wscore_vw, breaks = c(-Inf, 0, 1, 4.5, Inf), labels = c("<0", "0", "1-4", ">=5"), right = FALSE))
  }

  ### Check output for possible unknown-state errors
  .check_output(x = x, id = id, score = score)

  ### Factorise comorbidities if requested
  if (factorise) x <- .factorise(x = x, score = score)

  ### Label variables for RStudio viewer if requested
  if (labelled) x <- .labelled(x = x, score = score)

  ### Return a tidy data.frame
  return(x)
}
